### ISSUE TITLE
Option to save model with lowest cross-val loss

### ISSUE THREAD

**Issue by abuzarmahmood** [abuzarmahmood/blechRNN#13]

**Comment by abuzarmahmood** [comment]
### Summary of Suggestions and Changes

1. **Files Identified for Modification:**
   - **`train.py`:** Contains the `train_model` function responsible for the training process. It tracks cross-validation loss, and modifications are needed to save the model when the lowest loss is detected.
   - **`run_model.py`:** Responsible for running the model and saving it post-training. Adjustments are needed to incorporate the new functionality of saving the model with the lowest cross-validation loss.

2. **Plan to Address the Issue:**
   - Modify the `train_model` function in `train.py` to track and save the model state when a new lowest cross-validation loss is found.
   - Update the model saving logic in `run_model.py` to ensure the model with the lowest cross-validation loss is saved.

3. **Specific Changes Suggested:**

   - **Track and Save Model with Lowest Cross-Validation Loss:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/train.py`
     - **Lines:** 102-137
     - **Description:** Modify the `train_model` function to track the lowest cross-validation loss and save the model state when a new lowest loss is found.
     - **Code Snippet:**
       ```python
       def train_model(
               net, 
               inputs, 
               labels, 
               output_size,
               train_steps = 1000, 
               lr=0.01, 
               delta_loss = 0.01,
               device = None,
               criterion = MSELoss(), 
               test_inputs = None,
               test_labels = None,
               save_path = 'best_model.pt'  # New parameter for saving model
               ):
           ...
           best_cross_val_loss = float('inf')  # Initialize best loss
           for i in range(train_steps):
               ...
               if cross_val_bool and (i % 100 == 99): 
                   ...
                   if test_loss.item() < best_cross_val_loss:
                       best_cross_val_loss = test_loss.item()
                       torch.save(net.state_dict(), save_path)  # Save best model
           ...
       ```

   - **Adjust Model Saving Logic:**
     - **File:** `/home/exouser/Desktop/blech_github_bot/src/repos/abuzarmahmood/blechRNN/src/run_model.py`
     - **Lines:** 151-164
     - **Description:** Update the call to `train_model` to include the new `save_path` parameter for saving the best model.
     - **Code Snippet:**
       ```python
       net, loss, cross_val_loss = train_model(
               net, 
               train_inputs, 
               train_labels, 
               output_size = output_size,
               lr = 0.001, 
               train_steps = 15000,
               loss = loss_name, 
               test_inputs = test_inputs,
               test_labels = test_labels,
               save_path = os.path.join(artifacts_dir, f'best_{model_name}.pt')  # Specify save path
               )
       ```

These changes will ensure that the model with the lowest cross-validation loss is saved during the training process, effectively addressing the user's request.

---
*This response was automatically generated by blech_bot using model gpt-4o*

**Comment by abuzarmahmood** [comment]
Created pull request: https://github.com/abuzarmahmood/blechRNN/pull/14
Continue discussion there.

---
*This response was automatically generated by blech_bot using model gpt-4o*
